### PR TITLE
[mono] Add Vector128 AndNot intrinsic for amd64

### PR DESCRIPTION
**Pull request by matouskozak**
Add support for the following Vector128 API's:

- AndNot

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=74586) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=74586) before we can accept your contribution.<br/><br/>:x: matouskozak [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=74586)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=74586) it.</sub>

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -7950,6 +7950,10 @@ MONO_RESTORE_WARNING
 			case XBINOP_FORCEINT_XOR:
 				result = LLVMBuildXor (builder, lhs_int, rhs_int, "");
 				break;
+			case XBINOP_FORCEINT_ANDN:
+				rhs_int = LLVMBuildNot (builder, rhs_int, "");
+				result = LLVMBuildAnd (builder,  lhs_int, rhs_int, "");
```
Comment (tannergooding) [inline src/mono/mono/mini/mini-llvm.c:7955]: Does LLVM optimize this down to `andnps/andnpd` for float/double?

**Comment by tannergooding** [comment]
Could someone point out where the handling is that says this is valid for `V128` or not?

Likewise, could I get an explanation on why `V64`? x86/x64 doesn't have `V64` support (MMX "did" but is effectively deprecated and we don't expose it, it comes with a lot of net negatives to use this over just a software impl).

I'd expect `Vector128<T>` and `Vector256<T>` support for x86/x64. With `Vector128<T>` being always supported (part of the `SSE2` baseline) and `Vector256<T>` being supported when AVX2 is supported. Some APIs can optionally be accelerated when `AVX` is supported but `IsHardwareAccelerated` shouldn't report `true`.

I'd then expect `Vector64<T>` and `Vector128<T>` to be supported for Arm64. Both should be always supported (part of the `AdvSimd` baseline).

File: src/mono/mono/mini/mini-llvm.c (hunk 1 above)
Comment (matouskozak) [reply src/mono/mono/mini/mini-llvm.c:7955]: Actually, there is slight problem with this which I'm not sure what is causing it. For bitwise operations (BitwiseAnd, AndNot, etc.) implemented in Mono, LLVM is emitting for Vector128<T> only single precision floating point instruction, in this case `vandnps`.  I think numerically the results should be correct even for integer inputs, but I doubt it is the most effective solution for the CPU. When I examined the LLVM IR generated for integer inputs I didn't see any conversion to floats. I opened an issue for this https://github.com/dotnet/runtime/issues/74661.

Comment (tannergooding) [reply src/mono/mono/mini/mini-llvm.c:7955]: It matters less on modern CPUs and LLVM will generally do the "right" thing so provided Mono isn't itself saying they are all float, it's probably fine.

On some older or lower power devices, these could actually have different execution units and therefore using `andps` for integrals can cause a pipeline delay.

**Comment by matouskozak** [comment]
> Could someone point out where the handling is that says this is valid for `V128` or not?
> 
> Likewise, could I get an explanation on why `V64`? x86/x64 doesn't have `V64` support (MMX "did" have a limited subset for `__m64`, but it is effectively deprecated and we don't expose it, it comes with a lot of net negatives to use this over just a software impl).
> 
> I'd expect `Vector128<T>` and `Vector256<T>` support for x86/x64. With `Vector128<T>` being always supported (part of the `SSE2` baseline) and `Vector256<T>` being supported when AVX2 is supported. Some APIs can optionally be accelerated when `AVX` is supported but `IsHardwareAccelerated` shouldn't report `true`.
> 
> I'd then expect `Vector64<T>` and `Vector128<T>` to be supported for Arm64. Both should be always supported (part of the `AdvSimd` baseline).

We decided that we will stop emitting HW instrinsics for `Vector64<T>`. I'll create a separate PR for that where I disable it.

**Comment by radical** [comment]
I'm looking at whether the correct jobs are running on CI. @fanyang-mono @naricc do you see any missing builds here? Any that don't need to run?

**Comment by radical** [comment]
By *missing* I mean, any CI jobs that you would expect to run but are getting skipped.

**Comment by fanyang-mono** [comment]
@radical They are currently _queued_. I would need to check later to confirm.

**Comment by matouskozak** [comment]
Removed the `Vector64` support in `SN_AndNot`. This enables us to re-use `OP_SSE_ANDN` code.

**Review by fanyang-mono — APPROVED** [review]

**Pull request merged**